### PR TITLE
add tool icon preload

### LINKED ISSUE #704: Toolbar icons are loaded lazily
We should consider preloading images not embedded in client code. Currently we use such an approach for toolbar icons. It is because icons can be delivered in extensions.

https://streamable.com/00m2r

### PR DESCRIPTION
**Pull request by JulianWielga**
fixes #704

### CODE REVIEW

File: ui/client/reducers/selectors/settings.ts [hunk 1]
```diff
@@ -1,9 +1,10 @@
...
 import {SettingsState} from "../settings"
 import {createSelector} from "reselect"
+import {ProcessDefinitionData} from "../../types"
 
 const getSettings = (state: RootState): SettingsState => state.settings
 
 export const getFeatureSettings = createSelector(getSettings, s => s.featuresSettings)
 export const getLoggedUser = createSelector(getSettings, s => s.loggedUser)
-export const getProcessDefinitionData = createSelector(getSettings, s => s.processDefinitionData || {})
+export const getProcessDefinitionData = createSelector(getSettings, s => s.processDefinitionData || {} as ProcessDefinitionData)
```
Comment (lciolecki) [inline ui/client/reducers/selectors/settings.ts:10]: How does work this `as ..`?

File: ui/client/components/Tool.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        <img src={icon} alt={"node icon"} className="toolIcon"/>
+        {label}
+      </div>
+    </div>
+  )
+}
+
+const preloadImage = memoize((href: string) => new Image().src = href)
+
+export function useToolIcon(node: PossibleNode) {
```
Comment (lciolecki) [inline ui/client/components/Tool.tsx:40]: Why not arrow function?

File: ui/client/reducers/selectors/settings.ts (hunk 1 above)
Comment (JulianWielga) [reply ui/client/reducers/selectors/settings.ts:10]: https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions

File: ui/client/components/ToolBox.tsx [hunk 3]
```diff
@@ -39,8 +39,7 @@ export default function ToolBox() {
     <div id="toolbox">
       <div>
         {nodesToAdd.map(nodeGroup => {
-          const label =
-            <span className={"group-label"} onClick={() => toggleGroup(nodeGroup)}>{nodeGroup.name}</span>
+          const label = <span className={"group-label"} onClick={() => toggleGroup(nodeGroup)}>{nodeGroup.name}</span>
```
Comment (lciolecki) [inline ui/client/components/ToolBox.tsx:42]: Now you create for feach nodesToAdd new function. It's not good. Please create somethig like this:

```
clickGroupHandler = (node) => (node) => {

}
```
And use it:
```
 onClick={toggleGroup(clickGroupHandler)}
```
More info you can find here: https://medium.com/@machnicki/handle-events-in-react-with-arrow-functions-ede88184bbb

File: ui/client/reducers/selectors/settings.ts (hunk 1 above)
Comment (lciolecki) [reply ui/client/reducers/selectors/settings.ts:10]: O nice! Thanks ;)

File: ui/client/components/ToolBox.tsx (hunk 3 above)
Comment (JulianWielga) [reply ui/client/components/ToolBox.tsx:42]: it's not in scope of my change.

**Comment by lciolecki** [comment]
Please add info in docs about fixes.

**Pull request merged**